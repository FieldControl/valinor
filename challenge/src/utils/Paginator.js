//noinspection JSUnresolvedVariable
import React, { Component } from 'react'
import range from 'lodash/range'
import PropTypes from 'prop-types'

class PaginatorContainer extends Component {
    static propTypes = {
        initialPage: props => {
            if (typeof props.initialPage !== 'number') {
                return new Error('initialPage should be a number')
            }
            if (props.initialPage < 1) {
                return new Error('initialPage should be greater than or equal to 1')
            }
            if (props.initialPage > props.numberOfPages) {
                return new Error(
                    `initialPage should be less than or equal to ${props.numberOfPages}`
                )
            }
        },
        numberOfPages: PropTypes.number.isRequired,
        onPageChange: PropTypes.func,
        children: PropTypes.func.isRequired
    }

    static defaultProps = { initialPage: 1 }

    state = { currentPage: this.props.initialPage }

    setPage = page => {
        if (typeof page !== 'number') {
            throw new Error('page should be a number')
        }
        if (page < 1) {
            throw new Error('page should be greater than or equal to 1')
        }
        if (page > this.props.numberOfPages) {
            throw new Error(
                `page should be less than or equal to ${this.props.numberOfPages}`
            )
        }
        this.setState({ currentPage: page }, () => {
            if (this.props.onPageChange) {
                this.props.onPageChange(page)
            }
        })
    }

    incrementPage = () => {
        if (this.state.currentPage === this.props.numberOfPages) {
            return
        }
        this.setPage(this.state.currentPage + 1)
    }

    decrementPage = () => {
        if (this.state.currentPage === 1) {
            return
        }
        this.setPage(this.state.currentPage - 1)
    }

    render() {
        return this.props.children({
            currentPage: this.state.currentPage,
            numberOfPages: this.props.numberOfPages,
            setPage: this.setPage,
            incrementPage: this.incrementPage,
            decrementPage: this.decrementPage
        })
    }
}

const Paginator = ({ numberOfPages, onPageChange, initialPage }) => {
    return (
        <PaginatorContainer initialPage={initialPage} numberOfPages={numberOfPages} onPageChange={onPageChange}>
            {({ currentPage, setPage, incrementPage, decrementPage }) => (
                <button className="ml-auto mr-auto btn btn-group">
                    <button className="btn btn-outline-primary" disabled={currentPage === 1} onClick={e => { e.preventDefault(); decrementPage() }}>
                        &laquo;
                    </button>
                    {range(currentPage - 3, currentPage + 3 + 1).map(page => {
                        if (page < 1 || page > numberOfPages) {
                            return null
                        }
                        return (
                            <button key={page} className={`btn ${page === currentPage ? 'btn-primary' : 'btn-outline-primary'}`} onClick={e => { e.preventDefault(); setPage(page) }}>
                                {page}
                            </button>
                        )
                    })}
                    <button className="btn btn-outline-primary" disabled={currentPage === numberOfPages} onClick={e => { e.preventDefault(); incrementPage() }}>
                        &raquo;
                    </button>
                </button>
            )}
        </PaginatorContainer>
    )
}
Paginator.propTypes = {
    numberOfPages: PropTypes.number.isRequired,
    onPageChange: PropTypes.func.isRequired,
    initialPage: PropTypes.number.isRequired
}

export default Paginator