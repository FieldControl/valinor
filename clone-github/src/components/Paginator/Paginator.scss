.paginator {
  box-sizing: border-box;
  position: relative;
  display: flex;
  justify-content: center;
  align-items: center;
  margin: 25px auto;
  list-style: none;
}
.paginator__item {
  position: relative;
  box-sizing: border-box;
  & a {
    box-sizing: border-box;
    border: 1px solid silver;
    font-size: 1rem;
    text-decoration: none;
    width: 25px;
    height: 25px;
    padding: 10px;
  }
}
.paginator__control {
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  width: 50px;
  height: 50px;
  position: relative;
  box-sizing: border-box;
  margin-right: 5px;
  margin-left: 5px;
  padding: 5px;
  background-color: transparent;
  & img {
    width: 13px;
    height: 13px;
  }
}
.paginator__prev {
  box-sizing: border-box;
  position: relative;
  transform: rotate(90deg);
}
.paginator__next {
  box-sizing: border-box;
  position: relative;
  transform: rotate(-90deg);
}
.active {
  background-color: rgb(12, 48, 211);
  color: white;
  border: 1px solid blue;
}
