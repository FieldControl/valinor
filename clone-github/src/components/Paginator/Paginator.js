import React from "react";
import { NavLink, useLocation } from "react-router-dom";
import arrow from "../../assets/arrow.png";

import "./Paginator.scss";
import PropTypes from "prop-types";

const Paginator = ({ click, links }) => {
  const { page } = useLocation();

  let arr = [];
  if (links) {
    for (let i = 0; i < links.length; i++) {
      let element = links[i];
      if (page >= 3 && element.rel === "first") {
        arr[0] = element;
      }

      if (element.rel === "page") {
        arr[2] = element;
      }
      if (element.rel === "prev" && page >= 2) {
        arr[1] = element;
      }
      if (element.rel === "next") {
        arr[3] = element;
      }
      if (element.rel === "last" && element.rel === "next") {
        arr[4] = element;
      }
    }
  }
  return (
    <ul className="paginator">
      {links
        ? arr
            .filter((p) => {
              return p.rel === "prev";
            })
            .map((p, index) => {
              return (
                <li className="paginator__control paginator__prev" key={index}>
                  <NavLink
                    to={{
                      pathname: `/`,
                      page: p.page,
                    }}
                    onClick={click}
                    activeClassName=""
                  >
                    <img src={arrow} alt="arrow prev" />
                  </NavLink>
                </li>
              );
            })
        : null}

      {links
        ? arr.map((p, index) => {
            return (
              <li className="paginator__item" key={index}>
                <NavLink
                  to={{
                    pathname: `/`,
                    page: p.page,
                  }}
                  onClick={click}
                  activeClassName={
                    page === p.page || (page === undefined && p.page === 1)
                      ? "active"
                      : null
                  }
                >
                  {p.page}
                </NavLink>
              </li>
            );
          })
        : null}
      {links
        ? arr
            .filter((p) => {
              return p.rel === "next";
            })
            .map((p, index) => {
              return (
                <li className="paginator__control paginator__next" key={index}>
                  <NavLink
                    to={{
                      pathname: `/`,
                      page: p.page,
                    }}
                    activeClassName=""
                    onClick={click}
                  >
                    <img src={arrow} alt="arrow next" />
                  </NavLink>
                </li>
              );
            })
        : null}
    </ul>
  );
};

Paginator.propTypes = {
  click: PropTypes.func,
  links: PropTypes.array,
};
export default Paginator;
