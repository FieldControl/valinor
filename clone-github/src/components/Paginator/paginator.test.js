import { render, cleanup } from "@testing-library/react";
import Paginator from "./Paginator.js";
import { createMemoryHistory } from "history";
import { Router } from "react-router-dom";

afterEach(cleanup);

test("should take a snapshoot", () => {
  const history = createMemoryHistory();

  history.push("/");

  const { asFragment } = render(
    <Router history={history}>
      <Paginator />
    </Router>
  );
  expect(asFragment(<Paginator />)).toMatchSnapshot();
});
