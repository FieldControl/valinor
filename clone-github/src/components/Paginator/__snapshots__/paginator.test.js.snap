// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should take a snapshoot 1`] = `
<DocumentFragment>
  <ul
    class="paginator"
  />
</DocumentFragment>
`;
